proposals/TP2-dynamically-generated-cacheable-xds-resources.md

```diff
@@ -0,0 +1,1066 @@
+  }}
+]}
+```
+
+Now the deployment wants to add an additional key called `version`,
+whose value will be either `v1` or `v2`, so that it can further subdivide
+its clients' configs.
+
+If the new key is added on the clients first, then the clients will
+start subscribing with dynamic parameters constraints like the following:
+
+```textproto
+// For {env=prod, version=v1}
+{key_constraints:[
+  {key:"env" value:{
+    constraints:[{value:"prod"}]
+  }},
+  {key:"version" value:{
+    constraints:[{value:"v1"}]
+  }}
+]}
+```
+
+The server or cache has to match that set of constraints against the
+existing sets of dynamic parameters, which do not specify the `version`
+key at all.
+
+Conversely, if the new key is added on the server side first, then the
+server will have resource variants with parameters like this:
+- `{env=prod, version=v1}`
+- `{env=prod, version=v2}`
+- `{env=test, version=v1}`
+- `{env=test, version=v2}`
+
+But at this point, the clients are continuing to subscribe without
+constraints on this new key.  So the server or cache needs to figure out
+(e.g.) which of the first two sets of constraints to use for constraints
+that require `env` to be `prod` but do not specify `version`.
+
+We address this transition scenario by allowing the set of constraints
+for a given key to match any resource variant that does not specify that
+key at all.  This allows constraints for new keys to be added on clients
+before the corresponding keys are added on the resources on the server, but
+it does introduce some additional ambiguity into the matching.  For example,
+let's say that the server has the following two variants of a resource:
+- `{env=prod}`
+- `{env=prod, version=v1}`
+
+Now consider what happens if a client subscribes with the following
+constraints:
+
+```textproto
+{key_constraints:[
+  {key:"env" value:{
+    constraints:[{value:"prod"}]
+  }},
+  {key:"version" value:{
+    constraints:[{value:"v1"}]
+  }}
+]}
+```
+
+These constraints can match either of the above variants of the resource.
+This situation can be avoided by establishing a best practice that all
+variants of a given resource must have the same set of keys.
+
+There is still a possible ambiguity that can occur if a server adds
+multiple variants of a new key that clients are not yet sending.
+For example, let's say that the server has the following two variants
+of a resource:
+- `{env=prod, version=v1}`
+- `{env=prod, version=v2}`
+
+Consider what happens if a client subscribes with the following constraints:
+
+```textproto
+{key_constraints:[
+  {key:"env" value:{
+    constraints:[{value:"prod"}]
+  }}
+]}
+```
+
+These constraints can match either variant of the above resource.
+This can be avoided by establishing a best practice of not adding multiple
+variants of a new parameter until clients are sending the new parameter.
+However, if this does happen, the cache implementation is free to pick
+one of the variants at random.
```

Comment:
Arguably we should suggest it would be preferable to make this deterministic for any given client and known server state, i.e. some stability property. It's still not great though, since a client reconnect to a different server will result in something different.

Reply:
Yeah, I thought about that, but I don't think it will do much good.  This same matching needs to be done not just on the authoritative server but also in caching proxies and in the cache on leaf clients.  And there is no guarantee that the caches will have the same set of resources as the authoritative server, so they generally won't be able to make the same choice anyway.

I think the only way to get deterministic behavior is to require users to deploy the change starting at the clients.